### PR TITLE
codeforces 161D sol

### PR DESCRIPTION
**Pull request by TruaShamu**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by TruaShamu** [comment]
proof that it works
[java](https://codeforces.com/contest/161/submission/141204075)

[cpp](https://codeforces.com/contest/161/submission/141204030)

File: solutions/cf-161D.mdx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Main {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:33]:
```suggestion
public class DistanceInTree {
```

File: solutions/cf-161D.mdx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+			adj[b].add(a);
+		}
+
+		dfs(1, 0);
+		System.out.println(ans);
+
+	}
+
+	static void dfs(int cur, int par) {
+		dp[cur][0] = 1; // The current node's distance from itself fis 0.
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:61]:
```suggestion
		dp[cur][0] = 1;  // The current node's distance from itself is 0.
```

Style guide

File: solutions/cf-161D.mdx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			for (int i = 1; i <= K; i++) {
+				dp[cur][i] += dp[next][i - 1];
+			}
+		}
+	}
+}
+
+int main() {
+	cin >> N >> K;
+	// Edges
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:130]:
```suggestion
```

File: solutions/cf-161D.mdx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	for (int i = 0; i < N - 1; i++) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:131]:
```suggestion
	for (int i = 1; i < N; i++) {
```
I mean this isn't really necessary but shorter code.

File: solutions/cf-161D.mdx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		int a, b;
+		cin >> a >> b;
+		a--; b--;
+		adj[a].push_back(b);
+		adj[b].push_back(a);
+	}
+	dfs(0, -1);
+	cout << ans;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:139]:
```suggestion
	cout << ans << "\n";
```

File: solutions/cf-161D.mdx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+			if (next != par) {
+				// Visit children.
+				dfs(next, cur);
+				// Case 1 answer.
+				ans = dp[cur][K];
+
+				/*
+				 * Case 2 answer is the number of paths starting from subtree of cur
+				 * and ending in subtree of the unprocessed child.
+				 */
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:72]:
```suggestion
				
				// Case 2 answer is the number of paths starting from subtree of cur
				// and ending in subtree of the unprocessed child.
```

File: solutions/cf-161D.mdx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+		if (next != par) {
+			// Visit children.
+			dfs(next, cur);
+			// Case 1 answer.
+			ans = dp[cur][K];
+
+			/*
+			 * Case 2 answer is the number of paths starting from subtree of cur
+			 * and ending in subtree of the unprocessed child.
+			 */
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:116]:
```suggestion
			// Case 2 answer is the number of paths starting from subtree of cur
			// and ending in subtree of the unprocessed child.
```

File: solutions/cf-161D.mdx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Main {
+	static int N, K;
+	static ArrayList<Integer> adj[];
+	static int dp[][]; // DP[i][j] = number of nodes of distance j from node i.
```
Comment (Aadit-Ambadkar) [inline solutions/cf-161D.mdx:36]:
```suggestion
	static int dp[][];  // DP[i][j] = number of nodes of distance j from node i.
```
also this is over 80 char

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/cf-161D.mdx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+author: Sofia Yang
+---
+
+[Official Editorial](https://codeforces.com/blog/entry/4097)
+
+We want to find the total number of length $K$ paths in the tree.
+
+For each node $a$, there are 2 cases.
+
+1. There are length $K$ paths from $a$ to node $b$, in the subtree of $a$.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:14]:
```suggestion
1. There is a path of length $K$ from $a$ to node $b$, in the subtree of $a$.
```

File: solutions/cf-161D.mdx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+2. Otherwise, there are paths from node $b$ and node $c$ (both in the subtree of $a$), that pass through $a$.
+For case 2, the added length of the paths from $b$ to $a$, and the path from $c$ to $a$, should be $K$.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:16]:
```suggestion
2. There is a path from node $b$ to node $c$ (both in the subtree of $a$) that passes through node $a$.
The sum of the lengths of the paths from $b$ to $a$, and from $c$ to $a$, should be $K$.
```
You should clarify what node $b$ and $c$ are...

File: solutions/cf-161D.mdx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	
+
+If $N$ is the number of nodes, $K$ is the desired distance, we can define a DP matrix $\texttt{DP[N][K]}$ where $\texttt{DP[i][j]}$ is the number of nodes that has distance $j$ downwards from node $i$ in the tree.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:19]:
```suggestion
If $N$ is the number of nodes and $K$ is the desired distance, we can define a DP array where $\texttt{DP[i][j]}$ is the number of nodes that has distance $j$ downwards from node $i$ in the tree. The answer is $\texttt{DP[N][M]}$.
```
Could you clarify what "downwards" means?

File: solutions/cf-161D.mdx [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+So we want to DFS through the tree, and for each node, calculate the answer for case 1 and case 2.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:21]:
```suggestion
So we want to run DFS on the tree, and for each node, calculate the answer for case 1 and case 2.
```

File: solutions/cf-161D.mdx [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+Then we want to update the DP values by adding in the distances from the children.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:23]:
```suggestion
Then we want to update the DP values by adding in the distances from the children.
```
Maybe combine this sentence with that previous???

File: solutions/cf-161D.mdx [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+		dfs(1, 0);
+		System.out.println(ans);
+
+	}
+
+	static void dfs(int cur, int par) {
+		dp[cur][0] = 1; // The current node's distance from itself fis 0.
+		for (int next : adj[cur]) {
+			if (next != par) {
+				// Visit children.
```
Comment (jessechoe10) [inline solutions/cf-161D.mdx:64]:
```suggestion
				// Visit child node.
```

**Review by jessechoe10 — COMMENTED** [review]
You should be more specific with your comments and your editorial.

**Review by ar492 — CHANGES_REQUESTED** [review]
The cpp solution gives WA even on the samples.

**Comment by TruaShamu** [comment]
Sorry, there's something wrong with my solution.
I'll rewrite and make a new PR when I'm done.

**Pull request closed without merging**